src/components/application_manager/src/application_impl.cc

```diff
@@ -506,6 +515,12 @@ void ApplicationImpl::StopStreaming(
   using namespace protocol_handler;
   LOG4CXX_AUTO_TRACE(logger_);
 
+  // since WakeUpStreaming() is called from another thread, it is possible that
+  // the stream will be restarted after we call SuspendStreaming() and before
+  // we call StopXxxStreaming(). To avoid such timing issue, make sure that
+  // we run SuspendStreaming() and StopXxxStreaming() atomically.
+  sync_primitives::AutoLock lock(streaming_stop_lock_);
```

Comment:
Indeed, this PR includes the initial fix of PR https://github.com/smartdevicelink/sdl_core/pull/1783. We needed that fix to run our internal testing.
Do you want to remove changes related to PR https://github.com/smartdevicelink/sdl_core/pull/1783 ? Or do you want me to include the latest fix of it? Please let me know.

Reply:
Changes should be removed from here, since they are not related to the feature.

Reply:
Sorry for my delay. I remembered that this is *not* part of https://github.com/smartdevicelink/sdl_core/pull/1783. This fixes an issue that we didn't hit until we introduce multiple transports.
I'm OK to split this part into separate PR, however I would like to mention that the new PR will have dependency on this one.

__Description:__
When stopping secondary transport during video streaming, Core sometimes notifies Nav.OnVideoDataStreaming(true) between Nav.OnVideoDataStreaming(false) and Nav.StopStream().

__Cause:__
`StopNaviService()` will be triggered by`ConnectionHandlerImpl::OnSecondaryTransportEnded()` from Transport Manager's thread, while `ApplicationImpl::WakeUpStremaing()` will be called on Protocol Handler's thread when a video packet arrives. When these happen at the same time, they cause a timing issue described above.

__Notes:__
I don't think we'll hit this issue with single transport; when the transport is disconnected, it results application unregistration and streaming is stopped because of it.
This timing issue is causing bad effect on our HMI so I would like to include it in the next release.

Reply:
Hmmm, considering it was an issue which appeared because of the multiple transports feature, I'm fine with it being included here.